### PR TITLE
Add support for Arrays of primitives

### LINKED ISSUE #1194: Support lists of primitives
Depends on realm/realm-core#1881 and needs API design.

See [so question](http://stackoverflow.com/questions/41920861/is-it-possible-to-save-custom-sized-array-of-int-arrays-as-field-of-realmobject) asking if possible, for a good user-written answer on a workaround.

### PR DESCRIPTION
**Pull request by nirinchev**
## Description

Fixes #1194
Related to https://github.com/realm/product/issues/59

## TODO
- [x] `IList<string>` + Tests
- [x] `IList<byte[]>` + Tests
- [x] `IList<DateTimeOffset>` + Tests
- [x] Tests for ThreadSafeReferences
- [x] Changelog entry
- [x] Notification tests
- [x] Unmanaged object tests
- [x] Test `Remove`
- [x] Test `Clear`
- [x] Test `Move`
- [x] Test invalid input

### CODE REVIEW

File: Realm/Realm/RealmList.cs [hunk 1]
```diff
@@ -41,10 +43,11 @@ namespace Realms
     [Preserve(AllMembers = true)]
     [EditorBrowsable(EditorBrowsableState.Never)]
     [SuppressMessage("StyleCop.CSharp.DocumentationRules", "SA1600:ElementsMustBeDocumented")]
-    public class RealmList<T> : RealmCollectionBase<T>, IList<T>, IDynamicMetaObjectProvider where T : RealmObject
+    public class RealmList<T> : RealmCollectionBase<T>, IList<T>, IDynamicMetaObjectProvider
     {
         private readonly Realm _realm;
         private readonly ListHandle _listHandle;
+        private readonly bool _isPrimitive = !typeof(RealmObject).IsAssignableFrom(typeof(T));
```
Comment (fealebenpae) [inline Realm/Realm/RealmList.cs]: This can be static.

**Comment by bmunkholm** [comment]
@nirinchev Are there other remaining tasks that should be listed under "TODO"?

**Comment by nirinchev** [comment]
That's the generic TODO we have on our PRs, which I didn't bother to remove. I will now :)

File: wrappers/src/list_cs.cpp [hunk 2]
```diff
@@ -174,8 +174,13 @@ REALM_EXPORT void list_add_timestamp_ticks(List* list, int64_t value, NativeExce
     
 REALM_EXPORT void list_add_string(List* list, uint16_t* value, size_t value_len, bool has_value, NativeException::Marshallable& ex)
 {
-    StringData str  = has_value ? Utf16StringAccessor(value, value_len) : StringData();
-    add(list, str, ex);
+    if (has_value) {
+        Utf16StringAccessor str(value, value_len);
+        add(list, (StringData)str, ex);
```
Comment (fealebenpae) [inline wrappers/src/list_cs.cpp]: Don't use C-style casts. Try `StringData(str)` or `static_cast<StringData>(str)`.

File: Realm/Realm/Helpers/Operator.cs [hunk 3]
```diff
@@ -72,7 +73,46 @@ static GenericOperator()
                 var input = Expression.Parameter(typeof(T), "input");
                 try
                 {
-                    return Expression.Lambda<Func<T, U>>(Expression.Convert(input, typeof(U)), input).Compile();
+                    Expression convertFrom = input;
+                    var typeOfT = typeof(T);
+                    var isTNullable = false;
+                    if (typeOfT.IsConstructedGenericType && typeOfT.GetGenericTypeDefinition() == typeof(Nullable<>))
```
Comment (fealebenpae) [inline Realm/Realm/Helpers/Operator.cs]: You could use `ReflectionExtensions.IsClosedGeneric` here.

File: wrappers/src/notifications_cs.hpp [hunk 4]
```diff
@@ -62,6 +64,15 @@ namespace realm {
         return -1;
     }
     
+    inline std::vector<size_t> get_indexes_vector(IndexSet indexSet)
```
Comment (fealebenpae) [inline wrappers/src/notifications_cs.hpp]: I think the argument type should be `const IndexSet&` to avoid copying.

**Comment by nirinchev** [comment]
This still crashes on .NET Core on Linux and Mac.

**Pull request merged**